bouter/free/__init__.py

```diff
@@ -1,6 +1,175 @@
+                   "f{:d}_vy".format(i_fish),
+                   "f{:d}_theta".format(i_fish),
+                   "f{:d}_vtheta".format(i_fish),
+               ] + ["f{:d}_theta_{:02d}".format(i_fish, i) for i in range(n_segments)]
+
+
+    def _fish_renames(self, i_fish, n_segments):
+        return dict(
+            {
+                "f{:d}_x".format(i_fish): "x",
+                "f{:d}_vx".format(i_fish): "vx",
+                "f{:d}_y".format(i_fish): "y",
+                "f{:d}_vy".format(i_fish): "vy",
+                "f{:d}_theta".format(i_fish): "theta",
+                "f{:d}_vtheta".format(i_fish): "vtheta",
+            },
+            **{
+                "f{:d}_theta_{:02d}".format(i_fish, i): "theta_{:02d}".format(i)
+                for i in range(n_segments)
+            }
+        )
+
+
+    def _rename_fish(self, df, i_fish, n_segments):
+        return df.filter(["t"] + self._fish_column_names(i_fish, n_segments)).rename(
+            columns=self._fish_renames(i_fish, n_segments)
+        )
+
+
+    def extract_bouts(
+        self,
+        max_interpolate=2,
+        window_size=7,
+        recalculate_vel=False,
+        median_vel=False,
+        scale=None,
+        threshold=1,
+        **kwargs
+    ):
+        """ Extracts all bouts from a freely-swimming tracking experiment
+
+        :param exp: the experiment object
+        :param max_interpolate: number of points to interpolate if surrounded by NaNs in tracking
+        :param scale: mm per pixel, recalculated by default
+        :param max_frames: the maximum numbers of frames to process, useful for debugging
+        :param threshold: velocity threshold in mm/s
+        :param min_duration: minimal number of frames for a bout
+        :param pad_before: number of frames that gets added before
+        :param pad_after: number of frames added after
+
+        :return: tuple: (list of single bout dataframes, list of boolean arrays marking if the
+         bout i follows bout i-1)
+        """
+
+        df = self.behavior_log
+
+        scale = scale or get_scale_mm(self)
+
+        dt = np.mean(np.diff(df.t[100:200]))
+
+        n_fish = self.n_fish
+        n_segments = self.n_tail_segments
+        dfint = df.interpolate("linear", limit=max_interpolate, limit_area="inside")
+        bouts = []
+        continuous = []
+        for i_fish in range(n_fish):
+            if recalculate_vel:
+                for thing in ["x", "y", "theta"]:
+                    dfint["f{}_v{}".format(i_fish, thing)] = np.r_[
+                        np.diff(dfint["f{}_{}".format(i_fish, thing)]), 0
+                    ]
+
+            vel2 = (
+                dfint["f{}_vx".format(i_fish)] ** 2 + dfint["f{}_vy".format(i_fish)] ** 2
+            ) * ((scale / dt) ** 2)
+            if median_vel:
+                vel2 = vel2.rolling(window=window_size, min_periods=1).median()
+            bout_locations, continuity = utilities.extract_segments_above_threshold(
```

Comment:
the part of this function before this line should be split into several cached functions: the velocity recalculation (see how the vigor calculation or the tail filling-out in embedded experiments works) and the bout extraction